### PR TITLE
[Trino] Adding explain query method

### PR DESCRIPTION
**Pull request by agl29**
## What changes were proposed in this pull request?

- Explain query method added

## How was this patch tested?

- Manually
- Unittest added

### CODE REVIEW

File: desktop/libs/notebook/src/notebook/connectors/trino.py [hunk 1]
```diff
@@ -319,6 +319,28 @@ def _get_columns(self, database, table):
     ]
 
 
+  @query_error_handler
+  def explain(self, notebook, snippet):
+    statement = snippet['statement'].rstrip(';')
+    snippet['statement'] = 'EXPLAIN ' + statement
```
Comment (Harshg999) [inline desktop/libs/notebook/src/notebook/connectors/trino.py:325]: Is this required?

File: desktop/libs/notebook/src/notebook/connectors/trino.py [hunk 2]
```diff
@@ -319,6 +319,28 @@ def _get_columns(self, database, table):
...
+      query_client = TrinoQuery(self.trino_request, 'USE ' + snippet['database'])
+      query_client.execute()
+      query_client = TrinoQuery(self.trino_request, 'EXPLAIN ' + statement)
+      result = query_client.execute()
+      for row in result.rows:
+        for _row in row:
+          explanation += _row
+
+    return {
+      'status': 0,
```
Comment (Harshg999) [inline desktop/libs/notebook/src/notebook/connectors/trino.py:338]: Is the UI depending on this status field?

**Review by Harshg999 — COMMENTED** [review]
Looks good, I've left few review comments.

File: desktop/libs/notebook/src/notebook/connectors/trino.py (hunk 2 above)
Comment (agl29) [reply desktop/libs/notebook/src/notebook/connectors/trino.py:338]: Yes

File: desktop/libs/notebook/src/notebook/connectors/trino.py [hunk 3]
```diff
@@ -319,6 +319,28 @@ def _get_columns(self, database, table):
...
+    explanation = ''
+
+    if statement:
+      query_client = TrinoQuery(self.trino_request, 'USE ' + snippet['database'])
+      query_client.execute()
+      query_client = TrinoQuery(self.trino_request, 'EXPLAIN ' + statement)
+      result = query_client.execute()
+      for row in result.rows:
+        for _row in row:
+          explanation += _row
```
[the remark this answers is not in the document]
Comment (agl29) [reply desktop/libs/notebook/src/notebook/connectors/trino.py:335]: Yes, Same as we have in Hive and Impala.

File: desktop/libs/notebook/src/notebook/connectors/trino.py (hunk 1 above)
Comment (agl29) [reply desktop/libs/notebook/src/notebook/connectors/trino.py]: modified

File: desktop/libs/notebook/src/notebook/connectors/trino.py (hunk 2 above)
Comment (wing2fly) [reply desktop/libs/notebook/src/notebook/connectors/trino.py:336]: Will the TrinoQuery.execute() throws any exception?
If any exception, should the status be updated to -1?
Take a look existing beeswax API examples: `response = {'status': -1, 'message': ''}` are defined at the beginning of menthod.

File: desktop/libs/notebook/src/notebook/connectors/trino.py [hunk 4]
```diff
@@ -321,6 +321,24 @@ def _get_columns(self, database, table):
...
+    explanation = ''
+
+    if statement:
+      query_client = TrinoQuery(self.trino_request, 'USE ' + snippet['database'])
+      query_client.execute()
+      query_client = TrinoQuery(self.trino_request, 'EXPLAIN ' + statement)
+      result = query_client.execute() 
+
+    return {
+      'status': 0,
```
Comment (wing2fly) [inline desktop/libs/notebook/src/notebook/connectors/trino.py:336]: Will the TrinoQuery.execute() throws any exception?
If any exception, should the status be updated to -1?
Take a look existing beeswax API examples: response = {'status': -1, 'message': ''} are defined at the beginning of menthod.

Comment (agl29) [reply desktop/libs/notebook/src/notebook/connectors/trino.py:340]: Added the try catch.
We need to send the {'status': 0} to capture the output in UI. 
For the ref https://github.com/cloudera/hue/blob/master/desktop/libs/notebook/src/notebook/connectors/hiveserver2.py#L591

File: desktop/libs/notebook/src/notebook/connectors/trino.py [hunk 5]
```diff
@@ -321,6 +321,28 @@ def _get_columns(self, database, table):
...
+
+    if statement:
+      try:
+        query_client = TrinoQuery(self.trino_request, 'USE ' + snippet['database'])
+        query_client.execute()
+        query_client = TrinoQuery(self.trino_request, 'EXPLAIN ' + statement)
+        result = query_client.execute()
+        explanation = result.rows
+      except Exception as e:
+        explanation = str(e)
```
Comment (Harshg999) [inline desktop/libs/notebook/src/notebook/connectors/trino.py:337]: Adding error in explanation?

Comment (agl29) [reply desktop/libs/notebook/src/notebook/connectors/trino.py:337]: Yes, that's what we are doing in the the Hive https://github.com/cloudera/hue/blob/master/desktop/libs/notebook/src/notebook/connectors/hiveserver2.py#L608

File: desktop/libs/notebook/src/notebook/connectors/trino_tests.py [hunk 6]
```diff
@@ -253,3 +253,35 @@ def test_get_select_query(self):
       self.trino_api._get_select_query(database, table),
       expected_statement
     )
+
+
+  def test_explain(self):
```
Comment (Harshg999) [inline desktop/libs/notebook/src/notebook/connectors/trino_tests.py:258]: Add a unit test for exception scenario also?

File: desktop/libs/notebook/src/notebook/connectors/trino.py [hunk 7]
```diff
@@ -321,6 +321,28 @@ def _get_columns(self, database, table):
...
+    statement = snippet['statement'].rstrip(';')
+    explanation = ''
+
+    if statement:
+      try:
+        query_client = TrinoQuery(self.trino_request, 'USE ' + snippet['database'])
+        query_client.execute()
+        query_client = TrinoQuery(self.trino_request, 'EXPLAIN ' + statement)
+        result = query_client.execute()
+        explanation = result.rows
```
Comment (Harshg999) [inline desktop/libs/notebook/src/notebook/connectors/trino.py:335]: Curious: Would this look cleaner?
```python
TrinoQuery(self.trino_request, 'USE ' + snippet['database']).execute()
result = TrinoQuery(self.trino_request, 'EXPLAIN ' + statement).execute()
```

**Review by Harshg999 — APPROVED** [review]
LGTM!

File: desktop/libs/notebook/src/notebook/connectors/trino.py (hunk 5 above)
Comment (Harshg999) [reply desktop/libs/notebook/src/notebook/connectors/trino.py:337]: @JohanAhlen Is UI expecting this for both editor v1 and editor v2?
Just a heads up, we need to improve all this flow then once we pick up this roadmap item. cc @agl29

Comment (JohanAhlen) [reply desktop/libs/notebook/src/notebook/connectors/trino.py:335]: It's a bit odd indeed, @agl29 did you try how the UI behave if you would return a 500 request with the error message instead?

File: desktop/libs/notebook/src/notebook/connectors/trino.py (hunk 4 above)
Comment (JohanAhlen) [reply desktop/libs/notebook/src/notebook/connectors/trino.py:338]: @wing2fly we should try to move away from using the status flag, today API users expect the proper http error code when something goes bad. That being said the UI might have some old code that checks for status == 0.

**Review by JohanAhlen — APPROVED** [review]
Nice! Good to improve the exception handing a bit. In some cases we shouldn't follow old patterns, test it case-by-case and perhaps add a TODO or comment in case it's too much work for the task at hand

**Pull request merged**